framework/files/system/oem/01_ros-rootfs.yaml

```diff
@@ -34,23 +34,34 @@ stages:
           /var/lib/cni
         PERSISTENT_STATE_BIND: "true"
   rootfs.before:
-    - name: "Pull data from provider"
+    - &datasource
+      name: "Pull data from provider"
       datasource:
-        providers: ["cdrom"]
+        providers: ["cdrom", "gcp", "openstack", "aws", "azure", "hetzner", "packet", "scaleway", "vultr", "digitalocean", "metaldata" ]
         path: "/oem"
+    # Ensures that if no userdata is found at this stage,
+    # we retry later and run the boot hook
+    # This is relevant only on first-boot
+    - if: '[ ! -f /oem/userdata ]'
+      files:
+      - path: /oem/userdata_load
   rootfs.after:
     - if: '[ ! -f /run/cos/recovery_mode ] && [ ! -f /run/cos/live_mode ]'
       name: "Grow persistent"
       layout:
         device:
           label: COS_PERSISTENT
         expand_partition:
-          size: 0
   network:
-    - name: "Pull data from provider (local)"
-      datasource:
-        providers: ["aws", "gcp", "openstack", "cdrom"]
-        path: "/oem"
+    - <<: *datasource
+    # Trigger /oem/userdata if was fetched late
+    # when network was available
+    # This is relevant only on first-boot
+    - if: '[ -f /oem/userdata ] && [ -f /oem/userdata_load ]'
+      commands:
+      - elemental cloud-init -s initramfs /oem/userdata
```

Comment:
This is odd... why calling initramfs stage during network stage?

I think it is wrong if userdata includes an initramfs stage that can be called at network stage, why not just make the userdata include a `network.after` stage?

Reply:
so the problem relies on the fact that the input is in standard cloud-init format, and not in our extended syntax.

When `yip` reads cloud-inits files in the standard syntax it encodes them in the `boot` stage for user/files/etc and `initramfs` for setting up the hostname. I do believe too that we should move this encoding simply to `network`, but on the other end that implies that the cloud-init content wouldn't run if there is no network (thus, it wouldn't allow us to setup networking)

Reply:
I'd rather open a separate issue to make `yip` consistent and encode everything under `boot`, but moving it to `network` would leave any network settings via cloud-init in the cold

Reply:
I've opened https://github.com/rancher-sandbox/cOS-toolkit/issues/1124 for further discussion, because it has different implications. IF for example, we don't encode the hostname into initramfs, any other stage that might depend on the hostname might react differently, as `boot` might trigger _too late_

Reply:
Oh I see this is even more convoluted that what I initially thought 😅 

IMHO the solution would be not using cloud-init format... I am not sure how feasible this is. Are we using our custom cloud-init formatted config or are we pulling it form somewhere else? I also understand that using cloud-init format can be beneficial to ease creating a community around it, using already known tools/syntaxes helps.

Reply:
That's the issue indeed.. it is pulled from outside as user input, so this is totally user-facing in a variety of ways..

For example, rancherd configuration is _embedded_ into the cloud init one, so the user input is one single file with all the config needed, that is system+rancherd. 

I'm indeed in favor of our syntax which is much more flexible.. but "forcing" our syntax might be too much from the user perspective, and os2 is mostly documented only around the standard syntax. 

On the other way around, scoping it entirely on the standard syntax might also alleviate us to not have to care to support the extended syntax via datasources, but rather just that available for who is customizing os2 from a system level.

Reply:
This solution has some caveats in any case too. Because _if_ datasources pulls very late and we specify an hostname from there, the hostname is set anyway lower in the stack..